src/OmniSharp.Roslyn.CSharp/Services/Completion/CompletionListBuilder_Sync.cs

```diff
@@ -110,7 +110,15 @@ internal static partial class CompletionListBuilder
                         }
                 }
 
-                var commitCharacters = BuildCommitCharacters(completion.Rules.CommitCharacterRules, isSuggestionMode, commitCharacterRuleCache, commitCharacterRuleBuilder);
+                var treatAsASuggestion = isSuggestionMode ||
+                    (
+                        // The user hasn't actually typed anything and completion provider does
+                        // not request the item be hard-selected.
+                        completion.Rules.MatchPriority != MatchPriority.Preselect &&
+                        typedSpan.Length == 0 &&
+                        completion.Rules.SelectionBehavior != CompletionItemSelectionBehavior.HardSelection
+                    );
```

Comment:
Could we extract this as a shared helper, rather than duplicating twice?

Reply:
@333fred extracted!